dali/operators/reader/gds_mem.cc

```diff
@@ -12,40 +12,283 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#include <map>
 #include <memory>
+#include <mutex>
 #include <utility>
 #include <vector>
 #include "dali/operators/reader/gds_mem.h"
+#include "dali/util/cufile_helper.h"
+#include "dali/core/math_util.h"
 #include "dali/core/spinlock.h"
 #include "dali/core/mm/pool_resource.h"
+#include "dali/core/mm/detail/align.h"
 #include "dali/core/mm/malloc_resource.h"
 #include "dali/core/mm/composite_resource.h"
 
 namespace dali {
+namespace gds {
+
+static size_t GetGDSChunkSizeEnv() {
+  if (char *env = getenv("DALI_GDS_CHUNK_SIZE")) {
+    size_t s = atoll(env);
+    DALI_ENFORCE(!is_pow2(s), "GDS chunk size must be a power of 2");
+    DALI_ENFORCE(s >= kGDSAlignment && s <= (16 << 20),
+      make_string("GDS chunk size must be a power of two between ", kGDSAlignment, " and 16 M"));
+    return s;
+  } else {
+    return 2 << 20;  // default - 2 MiB
+  }
+}
+
+size_t GetGDSChunkSize() {
+  static size_t chunk_size = GetGDSChunkSizeEnv();
+  return chunk_size;
+}
+
+void RegisterChunks(void *start, int chunks, size_t chunk_size) {
+  char *addr = static_cast<char *>(start);
+  char *end = addr + chunks * chunk_size;
+  for (; addr < end; addr += chunk_size) {
+    CUDA_CALL(cuFileBufRegister(addr, chunk_size, 0));
+  }
+}
+
+void UnregisterChunks(void *start, int chunks, size_t chunk_size) {
+  char *addr = static_cast<char *>(start);
+  char *end = addr + chunks * chunk_size;
+  for (; addr < end; addr += chunk_size) {
+    CUDA_CALL(cuFileBufDeregister(addr));
+  }
+}
+
+///////////////////////////////////////////////////////////////////////////////
+// GDSRegisteredResource
+
+class GDSRegisteredResource : public mm::memory_resource<mm::memory_kind::device> {
+  void adjust_params(size_t &size, size_t &alignment) {
+    if (alignment < chunk_size_) {
+      alignment = chunk_size_;
+    }
+    size += alignment;  // we could go with alignment - 256, but that would risk having adjacent
+                        // blocks which might get glued in the pool - and that would lead to
+                        // failures in cudaMemcpy etc.
+  }
+
+  void *do_allocate(size_t size, size_t alignment) override {
+    adjust_params(size, alignment);
+    char *raw = nullptr;
+    CUDA_CALL(cudaMalloc(&raw, size));
+    char *aligned = mm::detail::align_ptr(raw, alignment);
+    size_t padding = aligned - raw;
+    size_t chunks = (size - padding) / chunk_size_;
+    assert(chunks * chunk_size_ >= size);
```

Comment:
Protection against overflow?

Reply:
Rather checking that the padding didn't eat up too much.

Reply:
Understood.